### PR TITLE
fix: Change STIGMAN_JWT_SERVICENAME_CLAIM to conform w/ standard. Don't create null user if no username claim found.

### PR DESCRIPTION
**Pull request by cd-rite**
Keycloak was previously using `clientId` but now uses `client_id` which conforms w/ oidc spec. This caused our fallback service name to fail with new versions of Keycloak. When service name is not specified, app will now fall back through azp, client_id, and clientId claims. 
Changed processing to throw a privilege error If no username can be resolved from token claims. 
Added an "Authorize Error" object to SmError that returns status 401.

### CODE REVIEW

File: api/source/utils/auth.js [hunk 1]
```diff
@@ -26,11 +27,23 @@ const verifyRequest = async function (req, requiredScopes, securityDefinition) {
...
-            displayName: decoded[config.oauth.claims.name] || decoded[config.oauth.claims.username] || decoded[config.oauth.claims.servicename] || 'USER',
             email: decoded[config.oauth.claims.email] ||  'None Provided'
+        }        
+        // Get username from configured claim in token, or fall back through precedence list
+        const usernamePrecedence = [config.oauth.claims.username, config.oauth.claims.servicename, "azp", "client_id", "clientId"]
+        for (const precedence of usernamePrecedence) {
+            if (decoded[precedence]) {
+                req.userObject.username = decoded[precedence];
+                break;
+            }
```
Comment (csmig) [inline api/source/utils/auth.js]: This block could be shortened to:
```
const req.userObject.username = decoded[usernamePrecedence.find(element => !!decoded[element])]
```

File: api/source/utils/auth.js [hunk 2]
```diff
@@ -26,11 +27,23 @@ const verifyRequest = async function (req, requiredScopes, securityDefinition) {
...
         }
-
+        if (req.userObject.username === undefined) {
+            throw(new SmError.PrivilegeError("No token claim mappable to username found"))
+        }    
+        // Get display name from specified claim in token, or use username
+        req.userObject.displayName = decoded[config.oauth.claims.name] || req.userObject.username
+        req.userObject.email = decoded[config.oauth.claims.email] ||  'None Provided'
```
Comment (csmig) [inline api/source/utils/auth.js]: Email value was already assigned by line 30

File: api/source/utils/config.js [hunk 3]
```diff
@@ -76,7 +76,7 @@ let config = {
         claims: {
             scope: process.env.STIGMAN_JWT_SCOPE_CLAIM || "scope",
             username: process.env.STIGMAN_JWT_USERNAME_CLAIM || "preferred_username",
```
Comment (csmig) [inline api/source/utils/config.js]: Any reason not to remove the `|| "preferred_username"` to be consistent with the change made to the next line for servicename?

**Review by csmig — COMMENTED** [review]
Logic is sound, suggested some changes, see individual comments

File: api/source/utils/config.js (hunk 3 above)
Comment (cd-rite) [reply api/source/utils/config.js]: agree

File: api/source/utils/auth.js (hunk 1 above)
Comment (cd-rite) [reply api/source/utils/auth.js]: agree

File: api/source/utils/auth.js (hunk 2 above)
Comment (cd-rite) [reply api/source/utils/auth.js]: agree

**Pull request merged**